src/Plugin/Field/FieldWidget/OptionalDateWidget.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Drupal\commerce\Plugin\Field\FieldWidget;
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\datetime\Plugin\Field\FieldType\DateTimeItem;
+use Drupal\datetime\Plugin\Field\FieldWidget\DateTimeDefaultWidget;
+
+/**
+ * Plugin implementation of the 'commerce_entity_select' widget.
+ *
+ * @FieldWidget(
+ *   id = "commerce_optional_date",
+ *   label = @Translation("Optional date"),
+ *   field_types = {
+ *     "datetime"
+ *   },
+ *   multiple_values = TRUE
+ * )
+ */
+class OptionalDateWidget extends DateTimeDefaultWidget {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $element = parent::formElement($items, $delta, $element, $form, $form_state);
+
+    switch ($this->getFieldSetting('datetime_type')) {
```

Comment:
Stylistic nitpick, do you need a switch when you have only two options? You could do $title = 'Provide a date' and an if that overrides it for datetime